braid/doc/braid.1
.\" Man page generated from reStructeredText.
.
.TH BRAID 1 "2013-02-11" "0.2.6" "Compilers and Programming Languages"
.SH NAME
BRAID \- Next-generation high-performance language interoperability
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" -*- rst -*-
.
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B braid [\-h] [\-\-gen\-sexp] [\-\-gen\-sidl] [\-c <language>] [\-s <language>]
[\-\-makefile] [\-i] [\-\-suppress\-contracts] [\-m <prefix>]
[\-\-debug] [\-\-profile] [\-\-version] [\-\-license] [\-v]
[<file.sidl> [<file.sidl> ...]]
.UNINDENT
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
Braid 0.2.6 \- Do magically wonderful things with SIDL (scientific interface
definition language) files. BRAID is a high\-performance language
interoperability tool that generates Babel\-compatible bindings for the Chapel
programming language. For details on using the command\-line tool, please
consult the BRAID man page and the Babel user\(aqs guide.
.SS About BRAID and its relation to Babel
[image: logos/braid-logo-small.png]
.sp
BRAID, the Braid system for Rewriting Abstract Descriptions, is a
compiler\-like tool to generate the glue code necessary for language
interoperability. Developed by the Components project [1] at Lawrence
Livermore National Laboratory, BRAID supports the Scientific Interface
Definition Language (SIDL) for the language\-independent declaration of
interfaces associated with scientific software packages.
.sp
The language bindings generated by BRAID are binary\-compatible to
those generated by the related Babel [2] tool. While Babel handles
traditional programming languages (such as C, C++, Fortran (77\-2008),
Python, and Java), BRAID adds support for modern PGAS [3] languages.
.sp
In practice, this means that if you want to, for example, let a Chapel
program use a component implemented in Fortran 95, you would invoke
Babel for the Fortran code and Braid for Chapel code:
.sp
.nf
.ft C
braid \-\-makefile \-\-client=Chapel interface.sidl
babel \-\-makefile \-\-server=F95    interface.sidl
.ft P
.fi
.IP [1] 5
\fI\%https://computation.llnl.gov/casc/components/\fP
.IP [2] 5
\fI\%https://computation.llnl.gov/casc/components/\fP
.IP [3] 5
\fI\%http://en.wikipedia.org/wiki/Partitioned_global_address_space\fP
.SS Supported Languages
.sp
At this moment Chapel client code is well\-tested and considered
stable. Chapel server works also fairly well, however, the Chapel
compiler imposes the limitation that there can only be one copy
of the Chapel runtime system per executable (e.g., 1 Chapel library
\fIor\fP one Chapel \fBmain()\fP function).
.TS
center;
|l|l|l|.
_
T{
Language
T}	T{
Client
T}	T{
Server
T}
_
T{
Chapel
T}	T{
stable
T}	T{
stable
T}
_
T{
UPC
T}	T{
testing
T}	T{
planned
T}
_
T{
X10
T}	T{
planned
T}	T{
planned
T}
_
.TE
.sp
In addition to all regular Babel/SIDL features, the Chapel backend
also extends the Chapel runtime to support \fIborrowed arrays\fP;
i.e., Arrays that are initialized by an external function and can be
passed to Chapel code without copying. It also provides an interface
for \fIdistributed arrays\fP which allow non\-Chapel code to access
non\-local data residing in the PGAS\-space.
.sp
The following features are not yet implemented:
.sp
Generic arrays, arrays of objects, structs, contracts, hooks and RMI.
.SS Installation
.sp
BRAID uses an autotools\-based build system, so the regular:
.sp
.nf
.ft C
\&./configure && make install
.ft P
.fi
.sp
will work just fine. Below is the step\-by\-step break\-down of the the
.sp
installation process:
.sp
.nf
.ft C
tar xvf braid\-0.2.3.tar.bz2 && cd braid\-0.2.3
mkdir build && cd build
\&../configure \-\-prefix=<INSTALL_PATH>
make [\-j<N>] && make install

# other useful commands
\&../configure \-\-help
make check
.ft P
.fi
.SS Compiling Chapel to run on SLURM\-managed Clusters with Infiniband
.sp
We found the following setup to be working with MPI (to spawn
processes) and direct Infiniband (for the actual communication).
.sp
It seems that Chapel (~1.5) cannot be compiled with a GCC version
newer than 4.6.X, because of issues with GASnet and the Chapel code
base. A known good version of GCC for compiling Chapel is 4.3.6.
To compile Chapel in CHPL_HOME use the following commands:
.sp
.nf
.ft C
export CHPL_COMM=gasnet
export CHPL_COMM_SUBSTRATE=ibv
make \-C ${CHPL_HOME}
.ft P
.fi
.sp
To initialize the environment (using bash) run:
.sp
.nf
.ft C
export CHPL_COMM=gasnet
export CHPL_COMM_SUBSTRATE=ibv
cd ${CHPL_HOME}
source util/setchplenv.bash
cd \-
.ft P
.fi
.sp
To run a Chapel program using Infiniband communication and MPI to spawn
new processes via SLURM:
.sp
.nf
.ft C
export MPIRUN_CMD="srun \-p<partition> \-N %N %P %A" # you can customize this
\&./program
.ft P
.fi
.sp
Those are substitutions made by Chapel (or GASnet), here is a list
taken from one of the Chapel\-internal Makefiles:
.sp
.nf
.ft C
# The following substitutions are performed:
#   %P = program executable name
#   %N = requested node count
#   %A = program arguments
#   %Q = program arguments w/ an extra level of quotes
#   %D = the current working directory
#   %H = hostfile (if any)
CONDUIT_RUNCMD = mpirun \-np %N %P %A
.ft P
.fi
.sp
The way I understand it SLURM provides the mpirun launcher on the
machines it manages. Chapel/GASnet uses this mechanism to trick SLURM
into launching processes on multiple nodes. If you compiled Chapel
with the \(aqibv\(aq conduit, however, no actual communication will be done
over MPI, instead GASnet will directly use the Infiniband network to
exchange data between nodes.
.sp
Using Infiniband directly is supposed to perform better because MPI
apparently does not do one\-sided communication well.
.SS Patching the Chapel compiler
.sp
You will need to patch your Chapel 1.4.0 compiler using the following
command _after_ running ./configure:
.sp
.nf
.ft C
bash <builddir>/patch\-chapel\-compiler.sh
.ft P
.fi
.sp
The script will patch and recompile the compiler automatically. You
can undo this operation by running:
.sp
.nf
.ft C
bash <builddir>/patch\-chapel\-compiler.sh \-\-undo
.ft P
.fi
.sp
this will reverse the effects of the patch.
.SS User\-visible dependencies
.sp
If you just want to compile and install BRAID, you will need:
\- Python:                Version 2.6 or higher
\- gcc, ld, Perl, AWK, sed
.sp
If you want to run the regression tests, you will also need:
\- GNU make               Version 3.74 or higher
\- Babel:                 Version 2.0 or higher
\- Chapel:                Version 1.4.0
\- Java:                  JVM 1.6 or higher
\- NumPy:                 Version 1.0.4 or higher
.SS Developer\-only dependencies
.INDENT 0.0
.IP \(bu 2
Make:          GNU make version 3.74 or higher
.IP \(bu 2
Autotools:     Version 2.65 or later
.IP \(bu 2
SWI\-Prolog:    Version 5.10.4 or higher (only needed if you intend to modify [ir,sidl].def)
.IP \(bu 2
Doxygen:       Version 1.6 or higher (disable with ./configure \-\-disable\-documentation)
.IP \(bu 2
graphviz:      (for Doxygen)
.IP \(bu 2
GNU flex
.IP \(bu 2
rst2man:       Contained in python\-docutils (Debian, Ubuntu, openSUSE and Fedora)
.UNINDENT
.SS User\(aqs Guide
.SS Anatomy of a Babel call
.sp
A call from one language into another starts with the client (the
caller) invoking a \fIstub\fP, which accepts all the arguments of the
method in the native data format (such as value representation or
memory layout) of the client language. The stub is automatically
generated by Braid/Babel and is usually output in the client\(aqs
language, or in C, if the native language is not expressive enough to
perform the argument conversion, which often involves byte\-level
memory manipulation. Stubs are very small and the compiler can
sometimes inline them. The stub converts the arguments into the
intermediate object representation (IOR) which is Babel\(aqs native data
format. It then calls the server implementation, which also has to
accept the IOR. On the server side resides a \fIskeleton\fP, which does
the reverse operation of converting the IOR into the native
representation of the server (callee). The skeleton invokes the
\fIimplementation\fP which is the actual method to be called. Upon return,
it converts all outgoing arguments and the return value into the IOR
and returns to the Stub, which performs the translation back to the
client\(aqs data format.
.SS Scientific Interface Definition Language (SIDL)
.sp
Braid cannot (yet) parse source code. In order to generate bindings
for a component the user needs to specify the interface that the
component provides in the SIDL format. A lengthy description of all
SIDL features can be found in the
.. _Babel Manual: \fI\%https://computation.llnl.gov/casc/components/docs/users_guide/\fP
, Chapter 6.
.SS Chapel Language Binding
.sp
This section describes how the various features of SIDL are mapped
onto Chapel constructs.
.SS Introduction
.sp
Chapel is a modern high\-level parallel programming language originally
developed by Cray Inc. as part of the DARPA HPCS program. In contrast
to traditional programming languages such as C++, the runtime system
of the language takes care of executing the code in parallel.  The
language still offers the user fine\-grained control over things such
as the data layout and the allocation of distributed data, but it
renders the tedious explicit encoding of communication through, e.g., a
message\-passing interface, obsolete.
.SS Basics
.sp
To see examples of how to use the Chapel language with Braid, please
refer to the many examples located in the \fBbraid/regression/interop\fP
directory. The corresponding SIDL files can be found in the Babel
sources the \fBregression\fP subdirectory.
.SS Static functions
.sp
Chapel classes cannot have static methods. BRAID will generate a
\fB<classname_>_static\fP module that contains all static methods
defined by the SIDL class \fIclassname\fP.
.SS Object lifecycle
.sp
Objects are created using the
\fB<classname_>_static.create(out exception)\fP
function. This indirection is necessary because constructors may throw
a SIDL exception. The
\fB<classname_>_static.wrap_<classname>(in obj, out exception)\fP
function can be used to create a Chapel version of an existing
externally created SIDL object.
.sp
Objects can be up\-casted by using the provided \fBas_<baseclass>()\fP
methods. Down\-casting can be down by using the generated
\fB<baseclass>_cast_<target>()\fP methods.
.SS Scalar datatypes
.sp
The following table lists the scalar types supported by SIDL and the
corresponding Chapel types used by the skeleton or stub while
converting Chapel code from or into the IOR.  The SIDL scalar types
are (with the exception of strings) of fixed length and were easy to
support especially since Chapel has parametric support for the number
of bits in the integral and floating point types which map to the same
representation as used by the IOR. It also has native types for both
single\-precision and double\-precision complex numbers and supports
\fIopaque\fP types that allow data to be passed around through Babel/BRAID
back into the original address space. Chapel also supports enumerated
type to defines a set of named constants.  On the other hand, the
Babel IOR and the Chapel compiler use different representations for
complex numbers, hence BRAID generates glue code to pass around
copies.  Since Chapel does not have a \fIchar\fP type, BRAID generates
code to convert Chapel unit\-length strings into chars using a
statically allocated lookup table.
.TS
center;
|l|l|l|.
_
T{
SIDL type
T}	T{
Size (in bits)
T}	T{
Corresponding Chapel type
T}
_
T{
bool
T}	T{
1
T}	T{
bool
T}
_
T{
char
T}	T{
8
T}	T{
string (length=1)
T}
_
T{
int
T}	T{
32
T}	T{
int(32)
T}
_
T{
long
T}	T{
64
T}	T{
int(64)
T}
_
T{
float
T}	T{
32
T}	T{
real(32)
T}
_
T{
double
T}	T{
64
T}	T{
real(64)
T}
_
T{
fcomplex
T}	T{
64
T}	T{
complex(64)
T}
_
T{
dcomplex
T}	T{
128
T}	T{
complex(128)
T}
_
T{
opaque
T}	T{
64
T}	T{
int(64)
T}
_
T{
string
T}	T{
varies
T}	T{
string
T}
_
T{
enum
T}	T{
32
T}	T{
enum
T}
_
.TE
.sp
Scalar Data Types in SIDL and their Chapel equivalents on a 64\-bit
machine.
.SS Array data types
.sp
SIDL arrays can be generated by using the \fBsidl.<type>_array.create()\fP
family of functions:
.sp
.nf
.ft C
var max_size = 1024;
var arr = sidl.double_array.create2dCol(max_size, max_size);
var a_meta = arr(1)
var a_data = arr(2)
for i in 0..#max_size do {
  for j in 0..#max_size do {
    a_data[i:int(32),j:int(32)] = 3.14;
  }
.ft P
.fi
.sp
The create functions return a tuple with SIDL\-related metadata and a
Chapel borrowed array. Borrowed arrays are an extension to Chapel;
they inherit all the syntactic sugar support Chapel provides
for natively defined arrays. Hence there is no change in the Chapel
code while using these arrays except during array creation. We require
a small extension to the Chapel compiler to support borrowed arrays.
Borrowed arrays have data blocks allocated external to the Chapel runtime
unlike traditional Chapel arrays where each array uses a data block
managed by the Chapel runtime. This avoids superfluous allocation and
deallocation of array data blocks while passing the array reference
between Babel/BRAID calls. It becomes the user\(aqs responsibility to manage
the memory while using borrowed arrays.
.sp
In the above example, a generic SIDL array<> version of the created
SIDL array could be accessed using the \fBa_meta.generic\fP data member.
Generic arrays are passed as Chapel \fBopaque\fP values. To cast them to
a proper sidl.Array, use the following code:
.sp
.nf
.ft C
proc generic_array_input(in ga: opaque /* array< > */) {
  var ga_meta = sidl.double_array.cast(ga);
  if ga_meta != nil then {
    var ga_data = createBorrowedArray1d(ga_meta);
    ga_data[0] = 42;
  }
}
.ft P
.fi
.sp
From the user\(aqs perspective there is no difference between Chapel
representation of SIDL arrays and R\-arrays. It should be noted that
non column\-major arrays will be automatically copied and converted in
the Chapel stub generated by BRAID.
.SS Exceptions
.sp
Chapel does not (yet) support exception handling. To throw an
exception, a Chapel method should create a new exception object and
assign it to the \fB_ex\fP argument before returning.
.SS Client\-side
.SS Implementation\-side
.SS UPC Language Binding
.sp
This section describes how the various features of SIDL are mapped
onto Unified Parallel C constructs.
.sp
Braid so far has been tested with Berkeley UPC v. 2.14.2.
.sp
TODO.
.SS Development status
.sp
BRAID is written in 98% Python; the SIDL scanner is implemented in
flex (C). Some of the Python sources are automatically generated from
a high\-level specification (sidl.def, ir.def) by a Prolog script. The
implementation language choice is motivated by Python being the
highest\-level language that we can assume to be pre\-installed on all
our target systems. So far we have three components:
.INDENT 0.0
.IP \(bu 2
A complete parser for SIDL which generates an object\-oriented
intermediate representation (IR)
.IP \(bu 2
A converter to an extensible s\-expression\-based language
independent IR
.IP \(bu 2
Code generators that convert this IR into Chapel and C code.
Other languages supported by Babel will follow.
.sp
To facilitate the writing of these code generators we put some
effort into extending Python with a pattern\-matching mechanism
for arbitrarily complex tuples. (And the s\-expressions from the
IR are internally represented as Python tuples.)
.UNINDENT
.sp
This diagram shows the work\-flow implemented in BRAID:
.sp
.nf
.ft C
             Parser               Conversion
+\-\-\-\-\-\-\-\-\-\-\-\-\-+  +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+  +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
| SIDL        |\-\-| SIDL\-based       |\-\-| Language indep. IR  |
|             |  | declarative IR   |  | (s\-expressions)     |
+\-\-\-\-\-\-\-\-\-\-\-\-\-+  +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+  +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
                                         |              |   |
                                         |   Code       |   |
                                         |   Generators |   |
                                         |              |   |
                                       +\-\-\-\-\-\-\-\-\-\-\-+ +\-\-\-\-\-\-\-\-\-\-+
                                       | Chapel    | | C, ...   |
                                       |           | |          |
                                       +\-\-\-\-\-\-\-\-\-\-\-+ +\-\-\-\-\-\-\-\-\-\-+
.ft P
.fi
.sp
The idea to use the s\-expression\-based IR to interface with ROTE
at some later point. Other components (e.g. PAUL) can also
generate this IR to get access to the code generators.
.SS Further Information
.sp
The following files are available at the top of the release directory
structure provide additional information on the Babel release:
.INDENT 0.0
.IP \(bu 2
BUGS:       Lists known bugs
.IP \(bu 2
COPYRIGHT:  Lawrence Livermore National Security, LLC notice
.IP \(bu 2
INSTALL:    Provides installation instructions
.IP \(bu 2
README:     This file
.sp
Additional background information can be found at our web site at
.sp
\fI\%http://www.llnl.gov/CASC/components/\fP
.sp
and
.sp
\fI\%http://compose\-hpc.sourceforge.net/\fP
.UNINDENT
.SH AUTHORS
.SS Former Members
.sp
Adrian Prantl: Architect, main author
\fI\%adrian@llnl.gov\fP, \fI\%adrian@complang.tuwien.ac.at\fP
.SS Summer Interns
.sp
Shams Imam, Rice University
.sp
Please report bugs to <\fI\%components@llnl.gov\fP>.
consult the BRAID man page and the Babel user\(aqs guide.
.SH OPTIONS
.SS positional arguments
.INDENT 0.0
.INDENT 3.5
<file.sidl>           SIDL files to use as input
.UNINDENT
.UNINDENT
.SS optional arguments
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B \-h,  \-\-help
show this help message and exit
.TP
.B \-\-gen\-sexp
generate an s\-expression
.TP
.B \-\-gen\-sidl
generate SIDL output again
.TP
.BI \-c \ <language>, \ \-\-client \ <language>
generate client code in the specified language
(Chapel, UPC, or any language supported through Babel)
.TP
.BI \-s \ <language>, \ \-\-server \ <language>
generate server code in the specified language
(Chapel, UPC, or any language supported through Babel)
.TP
.B \-\-makefile
generate a default GNUmakefile
.TP
.B \-i,  \-\-generate\-hooks
generate pre\-/post\-method hooks
.TP
.B \-\-suppress\-contracts
refrain from generating contract enforcement from SIDL
specs
.TP
.BI \-m \ <prefix>, \ \-\-make\-prefix \ <prefix>
<prefix> is prepended to the name of babel.make and
the symbols defined internally to allow Braid to be
run multiple times in a single directory.
.TP
.B \-\-debug
enable debugging features
.TP
.B \-\-profile
enable profiling
.TP
.B \-\-version
print version and exit
.TP
.B \-\-license
print licensing details
.TP
.B \-v,  \-\-verbose
print more debug info
.UNINDENT
.UNINDENT
.UNINDENT
.sp
Please report bugs to <\fI\%components@llnl.gov\fP>.
.SH COPYRIGHT NOTICE
.sp
Copyright (c) 2011\-2013 Lawrence Livermore National Security, LLC.
Produced at the Lawrence Livermore National Laboratory.
Written by the Components Team <\fI\%components@llnl.gov\fP>.
.sp
LLNL\-CODE\-473891.
All rights reserved.
.sp
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:
.INDENT 0.0
.IP \(bu 2
Redistributions of source code must retain the above copyright
notice, this list of conditions and the disclaimer below.
.IP \(bu 2
Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the disclaimer (as noted below)
in the documentation and/or other materials provided with the
distribution.
.IP \(bu 2
Neither the name of the LLNS/LLNL nor the names of its contributors
may be used to endorse or promote products derived from this
software without specific prior written permission.
.UNINDENT
.sp
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE
LIVERMORE NATIONAL SECURITY, LLC, THE U.S. DEPARTMENT OF ENERGY OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.sp
Additional BSD Notice
.INDENT 0.0
.IP 1. 3
This notice is required to be provided under our contract with the

braid/NEWS
-*- org -*-
* BRAID 0.2.6
** UPC backend
** Full struct support
** new tree pattern matcher
* SIDL object arrays
* BRAID 0.2.5
** Synched with Chapel 1.5
** Implemented contract checking
** Implemented generic arrays for Chapel
** Unified client and server generator code

** [module]_static.create_[module]() is now just [module]_static.create()
** Default index data type for SIDL arrays is now int(32), as it should be
** Patch for Chapel runtime initialization
** More robust Chapel library initialization
** Examples for parallel Chapel interoperability
* BRAID 0.2.4
** Synched with Chapel 1.4
** Included in Babel 2.0.0


braid/braid/braid.py
#!/usr/bin/env python
# -*- Python -*-
## @package braid
#
# Command line handling for the BRAID tool
#
# Please report bugs to <components@llnl.gov>.
#
# \authors <pre>
#
# Copyright (c) 2011, 2012 Lawrence Livermore National Security, LLC.
# Produced at the Lawrence Livermore National Laboratory
# Written by Adrian Prantl <adrian@llnl.gov>.
#  
# LLNL-CODE-473891.
# All rights reserved.
#  
# This file is part of BRAID. For details, see 
# http://compose-hpc.sourceforge.net/. 
# Please read the COPYRIGHT file for Our Notice and
# for the BSD License.
#
# </pre>
#
# \mainpage
# <h1>Welcome to Braid!</h1>
# (a.k.a. Babel 3)
#
# This is BRAID, the <b>B</b>RAID system for <b>r</b>ewriting
# <b>a</b>bstract <b>i</b>ntermediate <b>d</b>escriptions, a new tool
# that has a command line interface similar to that of Babel. BRAID is
# implemented in Python, making it very portable itself.  BRAID is a
# multi-faceted, term-based system for generating language
# interoperability glue code designed and developed as part of the
# COMPOSE-HPC project (http://compose-hpc.sourceforge.net/) to be a
# reusable component of software composability tools.
#
# From a user's perspective, BRAID is the tool that generates glue
# code for parallel PGAS languages, while Babel handles traditional
# HPC languages. Eventually, we intend to make this distinction
# invisible to the end user by launching both through the same front
# end.  Our Chapel language interoperability tool is the first of
# several applications envisioned for BRAID including optimized
# language bindings for a reduced set of languages and building
# multi-language interfaces to code without SIDL interface
# descriptions.
#
# The most important difference between BRAID and Babel is how the
# language backends are designed: In Babel each code generator is a
# fixed-function Java class that builds all the glue code out of
# strings. BRAID, on the other hand, creates glue code in a high-level
# <em>language-independent</em> intermediate representation (IR). This
# intermediate representation is then passed to a code generator which
# translates it into actual high-level code. At the moment there are
# code generators for C and Chapel, and also initial versions for
# Fortran, Java and Python. This architecture offers a higher
# flexibility than the static approach of Babel: For example,
# (object-)method calls in Babel need to be resolved by looking up the
# address of the method in a virtual function pointer table. Since
# Chapel has no means of dealing with function pointers (it implements
# its own object system instead), BRAID's Chapel code generator will
# generate a piece of C code to do the virtual function call <em>on
# the fly</em>, and place a static call to this helper function in
# lieu of the virtual function call.
# 
# Using this system we can reduce the number of times the language
# barrier is crossed to the minimum, leading to more code generated in
# the higher-level language, which again enables the compiler to do a
# better job at optimizing the program.
#
# Similar to Babel, BRAID can also be instructed to generate a
# <em>Makefile</em> that is used to compile both program and glue code
# and link them with any server libraries.  The Chapel compiler works
# by first translating the complete program into C and then invoking
# the system C compiler to create an executable binary. The Makefile
# created by BRAID intercepts this process after the C files have been
# generated and builds a
# <em>libtool</em>(http://www.gnu.org/software/libtool/) library
# instead. Libtool libraries contain both regular (<tt>.o</tt>) and
# position-independent (<tt>.so</tt>) versions of all the object
# files, which can be used for static and dynamic linking,
# respectively.
#
# The Chapel language already has basic support for interfacing with C
# code via the <tt>extern</tt> keyword. BRAID uses this interface as
# an entry point to open up the language for all the other languages
# supported by Babel.
#
# <h2>License</h2> 
# BRAID is released under the BSD License.
#
# <h2>Authors</h2> 
# Copyright (c) 2011, 2012, Lawrence Livermore National Security, LLC.<br/>
# Produced at the Lawrence Livermore National Laboratory.<br/>
# Written by Adrian Prantl <adrian@llnl.gov>.<br/>
# LLNL-CODE-473891. All rights reserved.<br/>
# <h3>Interns</h3>
# Shams Imam, Rice University (Summer 2010, 2011)
#
# <h2>Further Reading</h2>
#
# Adrian Prantl, Thomas Epperly, Shams Imam, Vivek Sarkar<br/>
# <a href="http://pgas11.rice.edu/papers/PrantlEtAl-Chapel-Interoperability-PGAS11.pdf">
# Interfacing Chapel with Traditional HPC Programming Languages</a><br/>
# <em>PGAS 2011: Fifth Conference on Partitioned Global Address Space
# Programming Models</em>, October 2011.
#
# Adrian Prantl, Thomas G.W. Epperly, Shams Imam<br/>
# <a href="http://doi.acm.org/10.1145/2148600.2148636/">
# Poster: connecting PGAS and traditional HPC languages.</a><br/>
# <em>SC Companion 2011</em>: 69-70, November 2011.
#


import argparse, re, sys
import config, legal

def braid(args):
    import chapel.chpl_backend as chpl_be
    import     upc.upc_backend as upc_be
    import sidl_parser, sidl_symbols, codegen

    # Babel pass-through?
    lang = args.client if args.client else args.server
    babel_langs = r'^(([cC]((\+\+)|(xx))?)|([fF]((77)|(90)|(03)))|([jJ]ava)|([pP]ython))$'
    if lang and re.match(babel_langs, lang):
        print "Invoking babel to handle language %s..." % lang
        import subprocess
        cmd = [config.BABEL_PREFIX+'/bin/babel']+sys.argv[1:]
        exit(subprocess.call(cmd))

    # argument post-processing
    args.gen_contracts = not args.suppress_contracts
    if not args.make_prefix:
        args.make_prefix = ''

    # No. Braid called to action!
    backend = []
    def select_backend(backend, lang):
       if backend: return
       if re.match(r'([cC]hapel)|(chpl)', lang):
           backend.append(chpl_be.GlueCodeGenerator(args))
       elif re.match(r'(UPC)|(upc)', lang):
           backend.append(upc_be.GlueCodeGenerator(args))
       else:
           print "**ERROR: (%s) Unknown language `%s'." % (sys.argv[0], args.client)
           exit(1)


    for sidl_file in args.sidl_files:
        sidl_ast = sidl_parser.parse(sidl_file)
        
        # SIDL operations
        if args.gen_sexp:
            print str(sidl_ast)
        if args.gen_sidl:
            print str(codegen.generate("SIDL", sidl_ast, args.debug))

        # Client code generation
        if args.client == None:
            pass
        else:
            select_backend(backend, args.client)
            sidl_ast = inject_sidl_runtime(sidl_ast, args)
            print "FIXME Handle imports here after injection of sidl runtime"
            sidl_ast, symtab = sidl_symbols.resolve(sidl_ast, args.verbose)
            backend[0].generate_bindings(sidl_file, sidl_ast, symtab, with_server=False)

        # Server code generation
        if args.server == None:
            pass
        else:
            select_backend(backend, args.server)
            sidl_ast = inject_sidl_runtime(sidl_ast, args)
            sidl_ast, symtab = sidl_symbols.resolve(sidl_ast, args.verbose)
            backend[0].generate_bindings(sidl_file, sidl_ast, symtab, with_server=True)

    if args.makefile:
        assert(backend)
        if args.client:
            backend[0].generate_makefile()
        else:
            backend[0].generate_makefile()

def inject_sidl_runtime(sidl_ast, args):
    """
    Parse the sidl, sidlx runtime library and inject it into the
    imports field of \c sidl_ast.
    """
    import sidl_parser
    if not config.HAVE_BABEL:
        print "**ERROR: Please reconfigure %s to have Babel support." \
              %config.PACKAGE_NAME
        exit(1)

    if args.verbose:
        print "loading library 'sidl'"
    _, _, _, sidl = sidl_parser.parse(config.SIDL_PATH+'/sidl.sidl')

    if args.verbose:
        print "loading library 'sidlx'"
    _, _, _, sidlx = sidl_parser.parse(config.SIDL_PATH+'/sidlx.sidl')

    # merge in the standard library
    new_imps = [sidl, sidlx]
    # preserve other imports for further processing
    sf, req, imps, defs = sidl_ast
    for loop_imp in imps:
        scope = loop_imp[1]
        name = '.'.join(scope[1])
        if (name not in ['sidl', 'sidlx']):
            new_imps.append(loop_imp)
    
    return sf, req, new_imps, defs




if __name__ == '__main__':
    cmdline = argparse.ArgumentParser(description=config.PACKAGE_STRING+'''
- Do magically wonderful things with SIDL (scientific interface
  definition language) files.

BRAID is a high-performance language interoperability tool that generates Babel-compatible bindings for the Chapel programming language. For details on using the command-line tool, please consult the BRAID man page and the Babel user's guide.
''',
    epilog="Please report bugs to <%s>."%config.PACKAGE_BUGREPORT)

    cmdline.add_argument('sidl_files', metavar='<file.sidl>', nargs='*',#type=file
			 help='SIDL files to use as input')

    cmdline.add_argument('--gen-sexp', action='store_true', dest='gen_sexp',
			 help='generate an s-expression')

    cmdline.add_argument('--gen-sidl', action='store_true', dest='gen_sidl',
			 help='generate SIDL output again')

    cmdline.add_argument('-c', '--client', metavar='<language>',
                         help='generate client code in the specified language'+
                         ' (Chapel, UPC, or any language supported through Babel)')

    cmdline.add_argument('-s', '--server', metavar='<language>',
                         help='generate server code in the specified language'+
                         ' (Chapel, UPC, or any language supported through Babel)')

    cmdline.add_argument('--makefile', action='store_true',
                         help='generate a default GNUmakefile')

    cmdline.add_argument('-i', '--generate-hooks', action='store_true', dest='gen_hooks',
                         help='generate pre-/post-method hooks')

    cmdline.add_argument('--suppress-contracts', action='store_true', dest='suppress_contracts',
                         help='refrain from generating contract enforcement from SIDL specs')

    #cmdline.add_argument('-f', '--fuse', action='store_true', dest='fuse',
    #                     help='experimental')

    cmdline.add_argument('-m', '--make-prefix', metavar='<prefix>', dest='make_prefix',
                         help='<prefix> is prepended to the name of babel.make and the '+
                         'symbols defined internally to allow Braid to be run multiple '+
                         'times in a single directory.')

    cmdline.add_argument('--debug',   action='store_true', help='enable debugging features')
    cmdline.add_argument('--profile', action='store_true', help='enable profiling')
    cmdline.add_argument('--version', action='store_true', help='print version and exit')
    cmdline.add_argument('--license', action='store_true', help='print licensing details')
    cmdline.add_argument('-v', '--verbose', action='store_true', help='print more debug info')    

    args = cmdline.parse_args()
    args.skip_rmi = False

    if args.license:
        print config.PACKAGE_STRING
        print '''
Copyright (c) 2010-2012 Lawrence Livermore National Security, LLC.
Produced at the Lawrence Livermore National Laboratory.
Written by Adrian Prantl <adrian@llnl.gov>.
LLNL-CODE-473891. All rights reserved.

For details, see {pkg_url}
'''.format(pkg_url=config.PACKAGE_URL)
        print legal.license
        exit(0)
    if args.version:
        print config.PACKAGE_STRING
        print '# This version of BRAID was configured on %s, %s.'%(
            config.BUILD_MACHINE,config.CONFIGURE_DATE)
        exit(0)
    if len(args.sidl_files) == 0:
        cmdline.print_help()
        exit(1)

    # Dependencies
    if args.makefile and not (args.client or args.server):
        print """
Warning: The option --makefile is only effective when used in
         conjunction with --client or --server!
"""

    if args.profile:
        # Profiling
        
        import cProfile
        cProfile.runctx('braid(args)', globals(), locals(), 'braid.prof')
        import pstats
        stats = pstats.Stats('braid.prof')

        stats.strip_dirs()
        stats.sort_stats('time', 'calls', 'cum')
        stats.print_stats(25)

        stats.sort_stats('cum')
        stats.print_stats(25)
    else:
        braid(args)

    exit(0)
